### PR TITLE
Show available variants first (#11605)

### PR DESCRIPTION
**Pull request by dbalabka**
| Q               | A
| --------------- | -----
| Branch?         | 1.7
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | #11605 
| License         | MIT

### CODE REVIEW

**Comment by dbalabka** [comment]
@lchrusciel could you please update if there is a chance to review these changes and get feedback?

**Comment by lchrusciel** [comment]
Hey @dbalabka I may reopen the issue and send it over to PO, but do not guarantee, that something will happen to it

**Comment by lchrusciel** [comment]
After reading #11605, I would say, that it is a matter of new filtration options on the variant list rather than a required change in our behavior. You have issues with your custom implementation, not with Sylius per see. I'm not convinced if it should land in the core. It complicates query and makes the business logic not obvious.

New filter (probably turned on by default) seems to be much better approach here

**Comment by dbalabka** [comment]
@lchrusciel I see what you mean. Unfortunately, it is complicated to implement any kind of advanced filtration using Doctrine and MySQL due to DQL limitations (e.g. there no [hierarchical queries](https://en.wikipedia.org/wiki/Hierarchical_and_recursive_queries_in_SQL) support). There are some workarounds but they provide poor DX and increase maintainability costs (e.g. [native SQL mapping](https://www.doctrine-project.org/projects/doctrine-orm/en/2.8/reference/native-sql.html) )

It would be great to have a simple and developer-friendly approach to customize product grids :)

**Comment by vvasiloi** [comment]
@dbalabka 
I understand 2 different problems from #11605.
One is resolving the default product variant, which could be solved with a new implementation of `Sylius\Component\Product\Resolver\ProductVariantResolverInterface`.
Another is sorting variants by availability in a grid.
The repository method you modified, by default, is only used in the shop's product grid (sylius_shop_product) which doesn't display variants.
Could you clarify it please?

**Comment by dbalabka** [comment]
@vvasiloi you probably right. I wasn't aware of the resolver and it might be a viable and easier solution. Still, most of the stores keeping unorderable products on the last pages by default. It will be useful to have such sorting.

**Comment by vvasiloi** [comment]
@dbalabka that is indeed very common and would be a useful addition to Sylius.
However, your proposed implementation doesn't do that.
It will only work when the products are sorted by price and will change which variant's price is used for sorting.
Before it used the first one and with your implementation it will use the first enabled.
Also, the queries for `$allVariants` and `$availableVariants` are identical, I believe you intended the first one to be without `->andWhere('v1.enabled = :enabled')`.
I will think about it and try to propose a better solution.

**Comment by vvasiloi** [comment]
@dbalabka do you want to continue with this or should I close it?

File: src/Sylius/Bundle/CoreBundle/Doctrine/ORM/ProductRepository.php [hunk 1]
```diff
@@ -95,24 +95,31 @@ public function createShopListQueryBuilder(
...
+                 ->andWhere('m1.id = :product_id')
+                 ->andWhere('v1.enabled = :enabled')
+                ->getDQL()
              ;
+            // ... show available variant if such exists
+            $availableVariants = $this->createQueryBuilder('m2')
+                ->select('min(v2.position)')
+                ->innerJoin('m2.variants', 'v2')
+                ->andWhere('m2.id = :product_id')
+                ->andWhere('v2.enabled = :enabled')
```
Comment (dbalabka) [inline src/Sylius/Bundle/CoreBundle/Doctrine/ORM/ProductRepository.php:110]: missing line `->andWhere('(v2.onHand - v2.onHold) > 0')`

**Comment by dbalabka** [comment]
@vvasiloi seems, there is not much interest in this feature from the community. Also, the solution looks very hacky. I don't think that it is worth adding additional complexity. IMO it is much easier to implement such complex sorting using Elastsearch instead of MySQL.

**Comment by jakubtobiasz** [comment]
Hi @dbalabka!
Thanks for your contribution, but we've decided we prefer to left is as a project-level decision. If we change our mind, we'll back here :).

**Comment by dbalabka** [comment]
@dbalabka , I respect your decision. Still, the problem is that overriding takes work, and the code quickly becomes messy. Please consider providing more convenient extension points.

**Pull request closed without merging**